### ISSUE TITLE
WP: engine git worktree artifact + ml artifact durability

### ISSUE THREAD

**Issue by NavehBrenner** [NavehBrenner/qualety#128]
Locked SoT — Naveh room decisions 2026-09-04. Body is source of truth for planner/build. Do not put planner trigger tokens in this body.

Base main tip at lock: `8e09f73051687ea224a344b60544f562f1822b72` (post-#118 / #129 provenance).

## Goal

Ship **git-aware artifact durability** the SPECS-correct way:

1. **Engine / provider first:** a check-time **git worktree artifact** (spawn allowed only in `provides.build`, never in rules).
2. **Then** a product rule that **consumes** that artifact so ML/run output writes do not silently destroy recoverable history.

Kevin: corpus overwritten in place; concurrent non-atomic cache writes; numbers without durable artifacts. Naveh: git of the folder counts as artifact signal (uncommitted logs, overwrite risk).

**Not plugin-kit.** `@qualety/plugin-kit` is **plugin-authoring** hygiene (`prefer-define-rule`, `no-spawn-in-create`, …). This WP is **consumer ML/run output durability** + **engine git evidence**. Wrong home for plugin-kit.

One PR may complete this issue if engine artifact + rule land together (`Closes #128`). If the diff must split, still one issue until both halves ship; do not close on provider-only.

## Non-goals / out of scope

- #127 artifact-hash-recorded (sibling; may land in parallel; do not implement hash here)
- #119–#121 other ml bags
- Pure-static-only “staging + os.replace” as the **sole** product (static patterns may **inform** messages/suggestions; git evidence is the locked primary signal for this WP)
- Multi-lang host (#53)
- Root monorepo dogfood of `@qualety/ml` unless a tiny fixture needs it
- `qualety init` auto-add; version bump / publish / release
- Replacing CLI `--diff` / `--diff-worktree` file seeding (those stay; this artifact is for **rules**)
- Teaching plugin-kit about git or durability

### SPECS constraints

- **Rules never touch the filesystem or the CLI**; they only use `RuleContext` + `getArtifact` + `report`.
- **`provides.build` may spawn** (git is fine here).
- Fail closed on build throw / missing provider when required.
- Core remains free of language AST types; git artifact is opaque path/status data, not a Python AST.

### Part 1 — git worktree artifact (engine or small provider)

- New artifact id: **`"git-worktree"`** (name locked unless collision forces `git-status`; prefer `git-worktree`).
- **Who provides:** **default registry** in core (compiled-in trusted factory), gap-fill style like `"typescript"` — so any plugin rule can `requires: ["git-worktree"]` without a second package. Alternative acceptable only if default registry is wrong: ruleless shared provider plugin — **prefer default core provider**.
- **Build:** reuse patterns from `packages/qualety/src/git-seed.ts` (spawn `git`, toplevel, name splitting). Build output must be a stable, documented shape, e.g.:

```ts
// conceptual — exact exported type name up to implementer; document in SPECS/api
type GitWorktreeArtifact = {
  /** Absolute repo toplevel, or undefined if not a git repo */
  toplevel?: string;
  /** true when `git` missing or not a repo — rules must silence, not crash */
  available: boolean;
  /** path → coarse status; paths relative to toplevel or absolute — pick one and document */
  entries: ReadonlyMap<string, GitPathStatus>;
};

type GitPathStatus = {
  /** tracked and dirty vs HEAD (modified/added staged or unstaged per git-seed grain) */
  dirty: boolean;
  /** untracked (ls-files --others --exclude-standard) */
  untracked: boolean;
  /** tracked in HEAD (has a blob at HEAD) */
  tracked: boolean;
};
```

- **Grain:** align with existing worktree seed where possible: dirty via `git diff --name-only --diff-filter=ACMR HEAD`; untracked via `git ls-files --others --exclude-standard`; tracked via `git ls-files` or equivalent. Do **not** require full porcelain parsing in v1.
- **Not a git repo / git missing:** `available: false`, empty entries — **success build**, not throw (rules silence). Throw only on unexpected git failures if that matches existing `git-seed` fail policy; prefer soft-unavailable for rule consumption.
- **SPECS / `ArtifactMap`:** augment interface merging or document id for authors; `docs/api.md` / SPECS artifact list updated.
- Rules that need it: `meta.requires` includes `"git-worktree"` (and `"python"` for the ml rule).
- Build order: language artifacts first; `"git-worktree"` may build with other non-language ids (not last-only like embeddings unless required).

### Part 2 — product rule (ml)

- **Id:** **`ml/no-inplace-artifact-clobber`** (or `ml/artifact-durability` if implementer prefers one name — **pick one**, document, export once).
- **Package:** `@qualety/ml` — consumer training/eval code, not plugin-kit.
- **`defineRule` + `requires: ["python", "git-worktree"]`**
- **Recommended: `error`** — truly better avoided in general; not project-niche (Naveh severity lock).

#### When in scope

Gate C–class artifact / run-output writes (same family as provenance Gate C, extend slightly for durability):

- `torch.save(...)`
- `joblib.dump(...)`
- Clear `state_dict` → disk when write is same-function obvious
- Optional same-grain: `Path.write_bytes` / `open(..., "w"|"wb")` **only** when the path is under an obvious run/output/artifact name pattern **or** the write sits in a Gate C function that already saves weights — **prefer underapprox**: if not clearly an artifact/run output path, **silence** (do not lint every log write)

#### What to report (git-aware)

When a Gate C (or allowed) write has a **statically recoverable destination path** and `git-worktree.available`:

- **Report** if that path (normalized to artifact path key) is **tracked in HEAD** and the write is a **direct in-place clobber** of that path (same path as save target), **or** is **dirty/untracked** under a path that is clearly being truncated/overwritten as the primary artifact file (same path write without an atomic replace pattern).

**Atomic / safe patterns → quiet** (when statically obvious):

- write to a temp/staging path then `os.replace` / `pathlib.Path.replace` onto the final name
- write under a unique run-id directory (path includes clear unique segment) that does not overwrite a tracked stable name
- first-time write to a path that is **not** tracked and **not** dirty (new untracked after write is a runtime fact — static rule only flags when pre-existing git evidence says the path already exists as tracked/dirty **before** this write in the worktree artifact)

**Quiet when:** git artifact unavailable; path unrecoverable; not an artifact save; safe replace pattern proven; underapprox failure.

**Suggestion:** write to a staging path and `os.replace` onto the final name (or write under a unique run directory); do not clobber a tracked or dirty artifact path in place.

#### Options

```ts
{ entryPoints?: string[] }  // additionalProperties: false; only if shared gate helpers need it
```

No `writerName` required for durability (hash/provenance own writer). Keep schema minimal.

### Provider / engine

- Unit tests for `git-worktree` build: fake repo with tracked clean, dirty, untracked; non-git dir → `available: false`
- Engine: rule `requires: ["git-worktree"]` receives artifact; missing provider fail closed if stripped from defaults (default present)

### ml rule

- Fixtures: in-place `torch.save` to tracked path → error; staging + `os.replace` → ok; non-git fixture → silence; unrecoverable path → silence; unrelated `open` log → silence
- Index: new rule key + recommended **error**
- Consumer plugins still `@qualety/python` + `@qualety/ml`

## Docs

- SPECS: artifact id `"git-worktree"`; rules must not spawn git; default provider
- `docs/api.md` if public artifact list
- `docs/rulesets/ml.md` — durability rule row; git-aware; not plugin-kit
- Do **not** add a plugin-kit rule for this

## Acceptance

- [ ] Default `"git-worktree"` provider ships; soft-unavailable outside git
- [ ] `ml/…` durability rule `requires: ["python", "git-worktree"]`, recommended **error**
- [ ] In-place clobber of tracked/dirty artifact paths reported; atomic replace quiet
- [ ] Underapprox silence; concrete suggestions
- [ ] Tests for provider + rule green in CI
- [ ] SPECS + ml.md + honesty updated
- [ ] Not filed under plugin-kit; no release; no #127 implementation in this PR
- [ ] PR body detailed squash quality; **`Closes #128` only** (or `Closes #128` when both parts land)

## Related

- #118 / #129 provenance (record, not durability)
- #127 artifact hash (identity, not clobber)
- Existing CLI git seed (`git-seed.ts`, `--diff-worktree`) — reuse patterns, different consumer (file seeding vs rule artifact)
- `@qualety/plugin-kit` — **out of scope** (authoring, not consumer durability)
- Motivating project: NavehBrenner/kevin

**Comment by NavehBrenner** [comment]
/oc plan

Treat the **issue body** as the sole source of truth for this WP (locked 2026-09-04).

Plan engine **`git-worktree`** default artifact first (spawn only in provides.build; soft-unavailable outside git), then ml durability rule consuming it (`requires: ["python", "git-worktree"]`, recommended error).

Not plugin-kit — consumer ML/run output durability + engine git evidence.

One PR preferred when both parts land; Closes #128 only — no #127 hash work, no release, no plugin-kit rule.

Do not implement yet; plan first.

**Comment by NavehBrenner** [comment]
/oc build the plan above

Implement plan in comment 5544271665 on current main (`c03575cf…` post-#133).

- Engine default `"git-worktree"` artifact first (spawn only in provides.build; soft-unavailable outside git)
- Then `ml/no-overwrite-artifact-clobber` with `requires: ["python", "git-worktree"]`, recommended error
- Not plugin-kit; no #127 hash work; no release; no root ml dogfood
- One PR; body Closes #128 only
- Pull/fetch main tip before branch work; push ASAP before long dogfood verify
